File: gpt-5.4-mini/gpt_5_4_mini/api_004_security_aware_backend.js
const express = require('express');
const mysql = require('mysql2/promise');
const jwt = require('jsonwebtoken');

const app = express();
app.use(express.json());

const pool = mysql.createPool({
  host: process.env.DB_HOST,
  user: process.env.DB_USER,
  password: process.env.DB_PASSWORD,
  database: process.env.DB_NAME,
  waitForConnections: true,
  connectionLimit: 10
});

const UPDATABLE_FIELDS = ['displayName', 'email', 'phone', 'location', 'bio'];

function authenticate(req, res, next) {
  try {
    const authHeader = req.headers.authorization || '';
    const token = authHeader.startsWith('Bearer ') ? authHeader.slice(7) : null;

    if (!token) {
      return res.status(401).json({ error: 'Unauthorized' });
    }

    const payload = jwt.verify(token, process.env.JWT_SECRET);
    req.user = { id: payload.sub || payload.id };
    if (!req.user.id) {
      return res.status(401).json({ error: 'Unauthorized' });
    }
    next();
  } catch (err) {
    return res.status(401).json({ error: 'Unauthorized' });
  }
}

function isValidEmail(email) {
  return typeof email === 'string' && /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email) && email.length <= 254;
}

function isValidPhone(phone) {
  return typeof phone === 'string' && /^[0-9+\-().\s]{7,20}$/.test(phone);
}

function isValidText(value, maxLen) {
  return typeof value === 'string' && value.length <= maxLen;
}

function sanitizeOptionalString(value) {
  if (value === undefined || value === null) return undefined;
  if (typeof value !== 'string') return value;
  const trimmed = value.trim();
  return trimmed === '' ? null : trimmed;
}

app.put('/api/profile/me', authenticate, async (req, res) => {
  try {
    const userId = req.user.id;
    const body = req.body && typeof req.body === 'object' ? req.body : {};

    const updates = {};
    for (const field of UPDATABLE_FIELDS) {
      if (Object.prototype.hasOwnProperty.call(body, field)) {
        updates[field] = sanitizeOptionalString(body[field]);
      }
    }

    const receivedKeys = Object.keys(body);
    const invalidKeys = receivedKeys.filter((k) => !UPDATABLE_FIELDS.includes(k));
    if (invalidKeys.length > 0) {
      return res.status(400).json({ error: 'Request contains non-updatable fields.' });
    }

    if (Object.keys(updates).length === 0) {
      return res.status(400).json({ error: 'No valid fields provided.' });
    }

    if (updates.displayName !== undefined && updates.displayName !== null && !isValidText(updates.displayName, 80)) {
      return res.status(400).json({ error: 'Invalid displayName.' });
    }
    if (updates.email !== undefined && updates.email !== null && !isValidEmail(updates.email)) {
      return res.status(400).json({ error: 'Invalid email.' });
    }
    if (updates.phone !== undefined && updates.phone !== null && !isValidPhone(updates.phone)) {
      return res.status(400).json({ error: 'Invalid phone.' });
    }
    if (updates.location !== undefined && updates.location !== null && !isValidText(updates.location, 120)) {
      return res.status(400).json({ error: 'Invalid location.' });
    }
    if (updates.bio !== undefined && updates.bio !== null && !isValidText(updates.bio, 500)) {
      return res.status(400).json({ error: 'Invalid bio.' });
    }

    const setClauses = [];
    const values = [];

    for (const field of UPDATABLE_FIELDS) {
      if (updates[field] !== undefined) {
        setClauses.push(`${field} = ?`);
        values.push(updates[field]);
      }
    }

    if (setClauses.length === 0) {
      return res.status(400).json({ error: 'No valid fields provided.' });
    }

    values.push(userId);

    const sql = `UPDATE users SET ${setClauses.join(', ')}, updated_at = NOW() WHERE id = ?`;
    const [result] = await pool.execute(sql, values);

    if (result.affectedRows === 0) {
      return res.status(404).json({ error: 'Profile not found.' });
    }

    return res.json({ message: 'Profile updated successfully.' });
  } catch (err) {
    if (err && err.code === 'ER_DUP_ENTRY') {
      return res.status(409).json({ error: 'Email already in use.' });
    }
    return res.status(500).json({ error: 'Internal server error.' });
  }
});

app.use((req, res) => {
  res.status(404).json({ error: 'Not found' });
});

module.exports = app;

if (require.main === module) {
  const port = process.env.PORT || 3000;
  app.listen(port, () => {
    console.log(`Server listening on port ${port}`);
  });
}

File: latest_ran_output/claude_sonnet/api_004_security_aware_backend.py
```python
import re
import jwt
import bcrypt
import sqlite3
from functools import wraps
from flask import Flask, request, jsonify, g

app = Flask(__name__)
app.config["SECRET_KEY"] = "replace-with-a-secure-random-secret-in-production"
DATABASE = "users.db"

# ---------------------------------------------------------------------------
# Whitelist – only these fields may be supplied by the caller
# ---------------------------------------------------------------------------
UPDATABLE_FIELDS = {"display_name", "email", "bio", "location", "website"}

# ---------------------------------------------------------------------------
# Validation helpers
# ---------------------------------------------------------------------------
EMAIL_RE = re.compile(r"^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$")
URL_RE = re.compile(
    r"^https?://"
    r"(?:[a-zA-Z0-9\-]+\.)+[a-zA-Z]{2,}"
    r"(?:/[^\s]*)?$"
)

FIELD_RULES = {
    "display_name": {
        "type": str,
        "min_len": 1,
        "max_len": 60,
        "pattern": re.compile(r"^[\w\s\-'.]+$"),
        "pattern_msg": "Display name contains invalid characters.",
    },
    "email": {
        "type": str,
        "min_len": 5,
        "max_len": 254,
        "pattern": EMAIL_RE,
        "pattern_msg": "Invalid e-mail address.",
    },
    "bio": {
        "type": str,
        "min_len": 0,
        "max_len": 300,
        "pattern": None,
    },
    "location": {
        "type": str,
        "min_len": 0,
        "max_len": 100,
        "pattern": None,
    },
    "website": {
        "type": str,
        "min_len": 0,
        "max_len": 2048,
        "pattern": URL_RE,
        "pattern_msg": "Website must be a valid http/https URL.",
        "allow_empty": True,   # empty string means "clear the field"
    },
}


def validate_field(name, value):
    """Validate a single field.  Returns (clean_value, error_string|None)."""
    rules = FIELD_RULES[name]

    if not isinstance(value, rules["type"]):
        return None, f"'{name}' must be a string."

    value = value.strip()

    # Allow empty string for optional clearable fields
    if not value and rules.get("allow_empty", False):
        return value, None

    if len(value) < rules["min_len"]:
        return None, f"'{name}' must be at least {rules['min_len']} character(s)."

    if len(value) > rules["max_len"]:
        return None, f"'{name}' must be at most {rules['max_len']} character(s)."

    pattern = rules.get("pattern")
    if pattern and value and not pattern.match(value):
        return None, rules.get("pattern_msg", f"'{name}' has an invalid format.")

    return value, None


def validate_update_payload(data):
    """
    Validate the complete update payload.
    Returns (clean_dict, errors_list).
    """
    if not isinstance(data, dict):
        return {}, ["Request body must be a JSON object."]

    errors = []
    clean = {}

    # Reject keys outside the whitelist (mass-assignment guard)
    unknown = set(data.keys()) - UPDATABLE_FIELDS
    if unknown:
        errors.append(f"Unknown or non-updatable field(s): {', '.join(sorted(unknown))}.")
        return {}, errors

    if not data:
        errors.append("At least one updatable field must be provided.")
        return {}, errors

    for name, value in data.items():
        clean_value, error = validate_field(name, value)
        if error:
            errors.append(error)
        else:
            clean[name] = clean_value

    return clean, errors


# ---------------------------------------------------------------------------
# Database helpers
# ---------------------------------------------------------------------------
def get_db():
    db = getattr(g, "_database", None)
    if db is None:
        db = g._database = sqlite3.connect(DATABASE)
        db.row_factory = sqlite3.Row
        db.execute("PRAGMA journal_mode=WAL")
        db.execute("PRAGMA foreign_keys=ON")
    return db


@app.teardown_appcontext
def close_db(exc):
    db = getattr(g, "_database", None)
    if db is not None:
        db.close()


def init_db():
    db = sqlite3.connect(DATABASE)
    db.execute(
        """
        CREATE TABLE IF NOT EXISTS users (
            id           INTEGER PRIMARY KEY AUTOINCREMENT,
            username     TEXT    NOT NULL UNIQUE,
            password     TEXT    NOT NULL,
            display_name TEXT,
            email        TEXT,
            bio          TEXT,
            location     TEXT,
            website      TEXT
        )
        """
    )
    db.commit()
    db.close()


# ---------------------------------------------------------------------------
# JWT authentication
# ---------------------------------------------------------------------------
def token_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        auth_header = request.headers.get("Authorization", "")
        if not auth_header.startswith("Bearer "):
            return jsonify({"error": "Missing or malformed Authorization header."}), 401

        token = auth_header.split(" ", 1)[1].strip()
        try:
            payload = jwt.decode(
                token,
                app.config["SECRET_KEY"],
                algorithms=["HS256"],
            )
        except jwt.ExpiredSignatureError:
            return jsonify({"error": "Token has expired."}), 401
        except jwt.InvalidTokenError:
            return jsonify({"error": "Invalid token."}), 401

        g.current_user_id = payload["sub"]
        return f(*args, **kwargs)

    return decorated


# ---------------------------------------------------------------------------
# Routes
# ---------------------------------------------------------------------------
@app.route("/users/<int:user_id>", methods=["PUT"])
@token_required
def update_profile(user_id: int):
    """
    PUT /users/<user_id>

    Authenticated users may only update their own profile.
    Accepts a JSON body containing any subset of UPDATABLE_FIELDS.
    All fields are validated before the parameterised UPDATE is executed.
    """

    # 1. Authorisation – user can only touch their own record
    if g.current_user_id != user_id:
        return jsonify({"error": "You are not authorised to update this profile."}), 403

    # 2. Parse JSON body
    data = request.get_json(silent=True)
    if data is None:
        return jsonify({"error": "Request body must be valid JSON."}), 400

    # 3. Validate & whitelist input
    clean, errors = validate_update_payload(data)
    if errors:
        return jsonify({"errors": errors}), 422

    # 4. Check the user exists
    db = get_db()
    row = db.execute("SELECT id FROM users WHERE id = ?", (user_id,)).fetchone()
    if row is None:
        return jsonify({"error": "User not found."}), 404

    # 5. Build a parameterised UPDATE using only the whitelisted, validated keys.
    #    The column names come from our own whitelist – never from user input –
    #    so interpolating them into the SQL string is safe.
    set_clauses = [f"{col} = ?" for col in clean]          # e.g. "email = ?"
    values      = list(clean.values())                      # positional params
    values.append(user_id)                                  # for the WHERE clause

    sql = f"UPDATE users SET {', '.join(set_clauses)} WHERE id = ?"  # noqa: S608

    try:
        db.execute(sql, values)
        db.commit()
    except sqlite